modules/tinymce/src/core/main/ts/dom/TrimNode.ts

```diff
@@ -46,33 +50,28 @@ const isDocument = (node: Node) => NodeType.isDocumentFragment(node) || NodeType
 //   <p>text 1<span></span></p><b>CHOP</b><p><span></span>text 2</p>
 // this function will then trim off empty edges and produce:
 //   <p>text 1</p><b>CHOP</b><p>text 2</p>
-const trimNode = <T extends Node>(dom: DOMUtils, node: T): T => {
+const trimNode = <T extends Node>(dom: DOMUtils, node: T, root?: Node): T => {
+  const rootNode = root || node;
   if (NodeType.isElement(node) && isBookmarkNode(node)) {
     return node;
   }
 
   const children = node.childNodes;
   for (let i = children.length - 1; i >= 0; i--) {
-    trimNode(dom, children[i]);
+    trimNode(dom, children[i], rootNode);
   }
 
-  if (!isDocument(node) && !isKeepTextNode(node) && Empty.isEmpty(SugarElement.fromDom(node), false)) {
-    if (NodeType.isElement(node)) {
-      const currentChildren = node.childNodes;
-
-      // Don't remove if there's a br as a child, eg: <p><br></p>
-      if (Arr.exists(currentChildren, NodeType.isBr)) {
-        return node;
-      }
-
-      // If the only child is a bookmark then move it up
-      if (currentChildren.length === 1 && isBookmarkNode(currentChildren[0])) {
-        node.parentNode.insertBefore(currentChildren[0], node);
-      }
+  // If the only child is a bookmark then move it up
+  if (NodeType.isElement(node)) {
+    const currentChildren = node.childNodes;
+    if (currentChildren.length === 1 && isBookmarkNode(currentChildren[0])) {
+      node.parentNode.insertBefore(currentChildren[0], node);
     }
+  }
 
+  // Remove any empty nodes
+  if (!isDocument(node) && !isContent(rootNode, node) && !isKeepElement(node) && !isKeepTextNode(node, rootNode)) {
```

Comment:
I know it's not really in scope for this PR, but swapping our functions between taking root, node and taking node, root is kinda confusing (and hard for typescript to protect us from at all).

Reply:
Yeah, I'll admit I got caught by this as well originally. I swapped the order around in `isContent`, as we normally use `root` last.